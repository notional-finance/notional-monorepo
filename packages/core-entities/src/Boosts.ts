import {
  Network,
  SECONDS_IN_DAY,
  floorToMidnight,
  getNowSeconds,
} from '@notional-finance/util';
import { TokenDefinition } from './Definitions';
import { TokenBalance } from './token-balance';
import { Registry } from './Registry';

export function getPointsPerDay(t: TokenBalance) {
  return (
    t.toUnderlying().abs().toFiat('USD').toFloat() *
    getArbBoosts(t.token, t.isNegative())
  );
}

export function getArbBoosts(b: TokenDefinition, isDebt: boolean) {
  if (b.network !== Network.arbitrum) return 0;

  if (b.tokenType === 'VaultShare') {
    // ezETH/wstETH
    if (b.vaultAddress === '0xd7c3dc1c36d19cf4e8cea4ea143a2f4458dd1937')
      return 0;
    return 8;
  } else if (
    b.tokenType === 'PrimeDebt' ||
    (b.tokenType === 'PrimeCash' && isDebt)
  ) {
    // Variable Borrow
    if (b.currencyId === 2) return 4; // DAI
    if (b.currencyId === 3) return 4; // USDC
    if (b.currencyId === 6) return 4; // FRAX
    if (b.currencyId === 8) return 4; // USDT
  } else if (b.tokenType === 'fCash' && isDebt) {
    // Fixed Borrow
    if (b.currencyId === 2) return 4; // DAI
    if (b.currencyId === 3) return 4; // USDC
    if (b.currencyId === 6) return 4; // FRAX
    if (b.currencyId === 8) return 4; // USDT
  } else if (b.tokenType === 'PrimeCash' && !isDebt) {
    // Variable Lend
    if (b.currencyId === 1) return 2; // ETH
    if (b.currencyId === 3) return 2; // USDC

    if (b.currencyId === 8) return 1; // USDT
    if (b.currencyId === 2) return 1; // DAI
    if (b.currencyId === 6) return 1; // FRAX
    if (b.currencyId === 5) return 1; // wstETH
  } else if (b.tokenType === 'fCash' && !isDebt) {
    // Fixed Lend
    if (b.currencyId === 1) return 2; // ETH
    if (b.currencyId === 3) return 2; // USDC

    if (b.currencyId === 8) return 1; // USDT
    if (b.currencyId === 2) return 1; // DAI
    if (b.currencyId === 6) return 1; // FRAX
    if (b.currencyId === 5) return 1; // wstETH
  }

  return 0;
}

export function getPointsAPY(
  boost: number,
  currentTotalPoints: number,
  totalARB: number,
  seasonStart: Date,
  seasonEnd: Date
) {
  const totalARBPerSeason = TokenBalance.fromFloat(
    totalARB,
    Registry.getTokenRegistry().getTokenBySymbol(Network.arbitrum, 'ARB')
  );

  const daysIntoSeason = Math.floor(
    (floorToMidnight(getNowSeconds()) - seasonStart.getTime() / 1000) /
      SECONDS_IN_DAY
  );
  const daysLeftInSeason = Math.ceil(
    Math.max(seasonEnd.getTime() / 1000 - floorToMidnight(getNowSeconds()), 0) /
      SECONDS_IN_DAY
  );
  const dailyEmissionRate = currentTotalPoints / daysIntoSeason;
  const projectedPointTotal =
    currentTotalPoints + dailyEmissionRate * daysLeftInSeason;
  const pointsPerDollar = boost;
  const arbSharePerDollar = pointsPerDollar / projectedPointTotal;

  const dailyArbPerDollarInUSD =
    totalARBPerSeason.toFiat('USD').toFloat() * arbSharePerDollar;

  const result = dailyArbPerDollarInUSD * 365 * 100

  // TODO: remove this when we get the season two points calculating correctly
  return result > 0 ? result : 0;
}
