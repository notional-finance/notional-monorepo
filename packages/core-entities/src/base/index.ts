export { BaseRegistry } from './base-registry';
