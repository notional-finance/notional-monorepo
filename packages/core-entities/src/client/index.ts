export { TokenRegistryClient } from './token-registry-client';
export { ConfigurationClient } from './configuration-client';
export { NOTERegistryClient } from './note-registry-client';
