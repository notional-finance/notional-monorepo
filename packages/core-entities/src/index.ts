export * from './token-balance';
export * from './Definitions';
export * from './Matchers';
export * from './Registry';
export { AccountFetchMode } from './client/account-registry-client';
export * from './exchanges';
export * from './vaults';
export { ConfigurationClient, NOTERegistryClient } from './client';
export * from './config/fiat-config';
export * from './server';
export { ClientRegistry } from './client/client-registry';
export type { ServerRegistry } from './server/server-registry';
export type { GraphDocument } from './server/analytics-server';
export {
  fetchGraph,
  loadGraphClientDeferred,
  fetchGraphPaginate,
} from './server/server-registry';
export * from './config/whitelisted-vaults';
export { DeprecatedVaults } from './server/vault-overrides';
export { SecondaryIncentiveToken } from './config/whitelisted-tokens';
export * from './Boosts';

export { NetworkClientModel, NetworkServerModel } from './models/NetworkModel';
