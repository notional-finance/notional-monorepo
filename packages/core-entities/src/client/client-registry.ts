import { Network } from '@notional-finance/util';
import crossFetch from 'cross-fetch';
import { BigNumber } from 'ethers';
import { CacheSchema } from '..';
import FixedPoint from '../exchanges/BalancerV2/fixed-point';
import { TokenBalance } from '../token-balance';
import { BaseRegistry } from '../base/base-registry';

const USE_CROSS_FETCH =
  process.env['NX_USE_CROSS_FETCH'] || process.env['NODE_ENV'] == 'test';

export abstract class ClientRegistry<T> extends BaseRegistry<T> {
  protected abstract cachePath(): string;

  constructor(public cacheHostname: string) {
    super();
  }

  public cacheURL(network: Network) {
    return `${this.cacheHostname}/${network}/${this.cachePath()}`;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected static reviver(key: string, value: any): any {
    if (value === undefined || value === null) {
      return value;
    } else if (Array.isArray(value)) {
      return value.map((v, i) =>
        ClientRegistry.reviver(`${key}:${i.toString()}`, v)
      );
    } else if (typeof value === 'object') {
      if (
        Object.prototype.hasOwnProperty.call(value, 'type') &&
        value.type === 'BigNumber'
      ) {
        return BigNumber.from(value);
      } else if (
        Object.prototype.hasOwnProperty.call(value, '_isFixedPoint') &&
        value._isFixedPoint
      ) {
        return FixedPoint.from(value._hex);
      } else if (
        Object.prototype.hasOwnProperty.call(value, '_isTokenBalance') &&
        value._isTokenBalance
      ) {
        return TokenBalance.fromJSON(value);
      }
    }

    return value;
  }

  protected async _fetch<T>(network: Network, urlSuffix?: string): Promise<T> {
    const _fetch = USE_CROSS_FETCH ? crossFetch : fetch;
    const cacheUrl = this.cacheURL(network);
    const result = await _fetch(
      urlSuffix ? `${cacheUrl}/${urlSuffix}` : cacheUrl
    );
    const body = await result.text();
    if (result.status !== 200) throw Error(`Failed Request: ${body}`);
    return JSON.parse(body, ClientRegistry.reviver);
  }

  protected async _refresh(
    network: Network,
    blockNumber?: number
  ): Promise<CacheSchema<T>> {
    return this._fetch(network, blockNumber?.toString());
  }
}
