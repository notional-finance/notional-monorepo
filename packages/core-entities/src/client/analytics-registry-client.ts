import {
  Network,
  RATE_PRECISION,
  ZERO_ADDRESS,
  getMidnightUTC,
  percentChange,
} from '@notional-finance/util';
import { Routes } from '../server';
import { ClientRegistry } from './client-registry';
import { Registry } from '../Registry';
import { BigNumber } from 'ethers';
import { OracleDefinition, PriceChange, TokenDefinition } from '../Definitions';
import {
  ASSET_PRICE_ORACLES,
  HistoricalOracles,
  VaultData,
} from '../server/analytics-server';
import { PRICE_ORACLES } from './oracle-registry-client';
import { TokenBalance } from '../token-balance';
import { FiatKeys } from '../config/fiat-config';

const APY_ORACLES = [
  'fCashOracleRate',
  'PrimeCashPremiumInterestRate',
  'PrimeDebtPremiumInterestRate',
  'nTokenBlendedInterestRate',
  'nTokenFeeRate',
  'nTokenIncentiveRate',
];

export class AnalyticsRegistryClient extends ClientRegistry<unknown> {
  protected cachePath() {
    return `${Routes.Analytics}/analytics`;
  }

  get USD() {
    return Registry.getTokenRegistry().getTokenBySymbol(Network.All, 'USD');
  }

  get ETH() {
    return Registry.getTokenRegistry().getTokenBySymbol(Network.All, 'ETH');
  }

  protected _getHistoricalOracles(network: Network) {
    return (super.getLatestFromSubject(network, 'historicalOracles') ||
      []) as HistoricalOracles;
  }

  /**
   * Returns a snapshot of PRICE_ORACLE data at a given timestamp in the past.
   * @param network for the given network
   * @param timestamp at this given timestamp, expected to be at midnight UTC
   * @returns OracleDefinition[] for use in the oracle-registry
   */
  getHistoricalOracles(network: Network, timestamp: number) {
    return this._getHistoricalOracles(network)
      .filter((o) => PRICE_ORACLES.includes(o.oracleType))
      .map((o) => {
        const { historicalRates, ..._o } = Object.assign({}, o);
        let latestRate = historicalRates.find((r) => r.timestamp === timestamp);
        if (!latestRate) {
          if (o.oracleType === 'fCashSettlementRate') {
            // If there is an fCash settlement rate, it only ever has a single
            // rate. The historical rate will not show up sometimes due to the
            // minTimestamp filter.
            latestRate = {
              rate: o.latestRate as string,
              blockNumber: 0,
              timestamp: timestamp,
              totalSupply: null,
              tvlUnderlying: null,
            };
          } else if (
            o.oracleType === 'Chainlink' &&
            o.base === ZERO_ADDRESS &&
            o.quote === ZERO_ADDRESS
          ) {
            // Fill in the base ETH rate
            latestRate = {
              rate: '1000000000000000000' as string,
              blockNumber: 0,
              timestamp: timestamp,
              totalSupply: null,
              tvlUnderlying: null,
            };
          } else {
            latestRate = {
              rate: '0' as string,
              blockNumber: 0,
              timestamp: timestamp,
              totalSupply: null,
              tvlUnderlying: null,
            };
          }
        }

        return Object.assign(_o, {
          latestRate: {
            rate: BigNumber.from(latestRate.rate),
            blockNumber: latestRate.blockNumber,
            timestamp: latestRate.timestamp,
          },
        }) as OracleDefinition;
      });
  }

  /**
   * Finds the matching token and returns the price history and TVL for the token
   * @param token a given asset token on Notional
   * @returns an array of the historical price and TVL in both underlying and USD
   */
  getPriceHistory(token: TokenDefinition) {
    const tokens = Registry.getTokenRegistry();
    const o = this._getHistoricalOracles(token.network).find(
      (o) => ASSET_PRICE_ORACLES.includes(o.oracleType) && o.quote === token.id
    );

    return (
      o?.historicalRates.map(({ timestamp, rate, tvlUnderlying: tvl }) => {
        const underlying = tokens.getTokenByID(token.network, o.base);
        const tvlUnderlying = tvl
          ? TokenBalance.from(BigNumber.from(tvl), token)
          : undefined;
        let tvlUSD: TokenBalance | undefined;

        try {
          tvlUSD = tvlUnderlying
            ? tvlUnderlying.toFiat('USD', timestamp)
            : undefined;
        } catch (e) {
          console.log(e);
        }
        return {
          timestamp,
          // Rate will be quoted in underlying terms
          priceInUnderlying: TokenBalance.from(
            BigNumber.from(rate),
            underlying
          ),
          tvlUSD,
          tvlUnderlying,
        };
      }) || []
    );
  }

  getHistoricalAPY(
    token: TokenDefinition
  ): { timestamp: number; totalAPY: number; [key: string]: number }[] {
    if (token.tokenType === 'VaultShare' && token.vaultAddress) {
      return this.getVault(token.network, token.vaultAddress).map(
        ({ timestamp, totalAPY }) => ({
          timestamp,
          totalAPY: totalAPY || 0,
        })
      );
    }

    const oracles = this._getHistoricalOracles(token.network).filter(
      (o) => o.quote === token.id && APY_ORACLES.includes(o.oracleType)
    );

    if (oracles.length === 1) {
      return oracles[0].historicalRates.map(({ timestamp, rate }) => {
        return {
          timestamp,
          totalAPY: (parseFloat(rate) / RATE_PRECISION) * 100,
        };
      });
    } else if (oracles.length > 1) {
      // Reduce across the multiple oracles and the given time frame and output all the
      // APY components
      const allTimestamps = oracles[0].historicalRates.map((h) => h.timestamp);
      const tokens = Registry.getTokenRegistry();

      return allTimestamps.map((t) => {
        const apyComponents = oracles.reduce(
          (acc, o) => {
            const r = o.historicalRates.find((h) => h.timestamp === t);
            let apy = 0;
            if (r) {
              apy = (parseFloat(r.rate) / RATE_PRECISION) * 100;

              if (o.oracleType === 'nTokenIncentiveRate') {
                apy = TokenBalance.fromFloat(
                  apy.toFixed(8),
                  tokens.getTokenBySymbol(token.network, 'NOTE')
                )
                  .toToken(tokens.getTokenByID(token.network, o.base))
                  .toFloat();
              }

              acc[o.oracleType] = apy;
            }
            acc['totalAPY'] += apy;
            return acc;
          },
          { totalAPY: 0 } as { totalAPY: number; [key: string]: number }
        );

        return {
          timestamp: t,
          ...apyComponents,
        };
      });
    } else {
      return [];
    }
  }

  getNTokenFeeRate(token: TokenDefinition) {
    const apyData = this.getHistoricalAPY(token);

    if (apyData.length === 0) return 0;
    else return apyData[apyData.length - 1]['nTokenFeeRate'] || 0;
  }

  getVault(network: Network, vaultAddress: string) {
    return (super.getLatestFromSubject(network, vaultAddress) ||
      []) as VaultData;
  }

  getPriceChanges(
    base: FiatKeys,
    network: Network,
    timeRange: number
  ): PriceChange[] {
    return Registry.getTokenRegistry()
      .getAllTokens(network)
      .filter((t) => t.currencyId !== undefined)
      .map((t) => {
        const unit = TokenBalance.unit(t);
        const midnight = getMidnightUTC();
        const pastDate = midnight - timeRange;
        const currentUnderlying = unit.toUnderlying();
        const currentFiat = unit.toFiat(base);
        let pastUnderlying: TokenBalance | undefined;
        let pastFiat: TokenBalance | undefined;
        try {
          pastUnderlying = unit.toUnderlying(pastDate);
          pastFiat = unit.toFiat(base, pastDate);
        } catch {
          pastUnderlying = undefined;
          pastFiat = undefined;
        }

        return {
          asset: t,
          pastDate: pastDate,
          currentUnderlying,
          currentFiat,
          pastUnderlying,
          pastFiat,
          fiatChange: percentChange(currentFiat.toFloat(), pastFiat?.toFloat()),
          underlyingChange: percentChange(
            currentUnderlying.toFloat(),
            pastUnderlying?.toFloat()
          ),
        };
      });
  }

  async getView<T>(network: Network, viewName: string) {
    return this._fetch<T[]>(network, viewName);
  }
}
