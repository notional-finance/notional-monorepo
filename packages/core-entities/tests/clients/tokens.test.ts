import {
  AssetType,
  encodeERC1155Id,
  Network,
  ZERO_ADDRESS,
} from '@notional-finance/util';
import {
  AccountFetchMode,
  Registry,
  TokenBalance,
  TokenDefinition,
} from '../../src';
import { OracleRegistryClient, TokenRegistryClient } from '../../src/client';
import { BigNumber } from 'ethers';

const ETH_PRICE = 1798.536;
describe.withRegistry(
  {
    network: Network.ArbitrumOne,
    fetchMode: AccountFetchMode.SINGLE_ACCOUNT_DIRECT,
  },
  'Token Registry',
  () => {
    let tokens: TokenRegistryClient;
    let oracles: OracleRegistryClient;

    beforeAll(async () => {
      tokens = Registry.getTokenRegistry();
      oracles = Registry.getOracleRegistry();
    });

    it('gets all tokens', () => {
      expect(tokens.getAllTokens(Network.ArbitrumOne).length).toBe(54);
    });

    it('finds token by address', () => {
      expect(
        tokens.getTokenByAddress(Network.ArbitrumOne, ZERO_ADDRESS)?.name
      ).toBe('Ether');
    });

    it('finds tokens by symbol', () => {
      expect(
        tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC')?.address
      ).toBe('0xff970a61a04b1ca14834a43f5de4533ebddb5cc8');
    });

    it('finds tokens by id', () => {
      const id = encodeERC1155Id(6, 1695168000, AssetType.FCASH_ASSET_TYPE);
      expect(tokens.getTokenByID(Network.ArbitrumOne, id).name).toBe(
        'fFRAX Maturing 1695168000'
      );
    });

    it('allows a custom token to be registered', () => {
      const t = tokens.getTokenBySymbol(Network.ArbitrumOne, 'wstETH')!;

      tokens.registerToken({
        ...t,
        symbol: 'PEPE',
        id: '0x123',
      });

      expect(tokens.getTokenByID(Network.ArbitrumOne, '0x123').symbol).toBe(
        'PEPE'
      );
    });

    it('parses an input into a token balance instance', () => {
      const tb = tokens.parseInputToTokenBalance(
        '1000',
        'USDC',
        Network.ArbitrumOne
      );
      expect(tb.toDisplayStringWithSymbol()).toBe('1,000.000 USDC');
    });

    describe('Token Balance', () => {
      it('throws errors on mismatches', () => {
        const tb1 = TokenBalance.zero(
          tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC')!
        );
        const tb2 = TokenBalance.zero(
          tokens.getTokenBySymbol(Network.ArbitrumOne, 'DAI')!
        );

        expect(() => tb1.eq(tb2)).toThrowError();
        expect(() => tb1.lt(tb2)).toThrowError();
        expect(() => tb1.lte(tb2)).toThrowError();
        expect(() => tb1.gt(tb2)).toThrowError();
        expect(() => tb1.gte(tb2)).toThrowError();
        expect(() => tb1.add(tb2)).toThrowError();
        expect(() => tb1.sub(tb2)).toThrowError();
      });

      it('properly formats tokens across options', () => {
        const tb1 = tokens.parseInputToTokenBalance(
          '0',
          'USDC',
          Network.ArbitrumOne
        );
        expect(tb1.toDisplayString()).toBe('0.000');
        // Negative zero is truncated
        expect(tb1.copy(-1).toDisplayString()).toBe('0.000');
        // Negative and positive numbers display properly
        expect(tb1.copy(-1.23e6).toDisplayString()).toBe('-1.230');
        expect(tb1.copy(1.23e6).toDisplayString()).toBe('1.230');

        // Abbreviations work
        expect(tb1.copy(123e6).toDisplayString(2, true)).toBe('123.00');
        expect(tb1.copy(123e9).toDisplayString(2, true)).toBe('123.00k');
        expect(
          tb1.copy(BigNumber.from(10).pow(12).mul(123)).toDisplayString(2, true)
        ).toBe('123.00m');
        expect(
          tb1.copy(BigNumber.from(10).pow(13).mul(123)).toDisplayString(2, true)
        ).toBe('1.23b');
      });

      it('can convert between tokens', () => {
        const eth = tokens.parseInputToTokenBalance(
          '1',
          'ETH',
          Network.ArbitrumOne
        );

        expect(
          eth
            .toToken(tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC'))
            .toFloat()
        ).toBeCloseTo(ETH_PRICE);
      });

      it('can convert prime cash to underlying tokens', () => {
        const peth = tokens.parseInputToTokenBalance(
          '1',
          'pEther',
          Network.ArbitrumOne
        );

        expect(peth.toUnderlying().toDisplayStringWithSymbol(8)).toBe(
          '1.00000000 ETH'
        );
      });
    });

    describe('Exchange Rates', () => {
      it('[FORWARD] can find a path from eth => usdc', () => {
        const eth = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH')!;
        const usdc = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC')!;
        const path = oracles.findPath(eth.id, usdc.id, Network.ArbitrumOne);
        expect(path).toEqual([eth.id, usdc.id]);
        const value = oracles.getLatestFromPath(Network.ArbitrumOne, path);
        expect(oracles.formatNumber(value!)).toBeCloseTo(ETH_PRICE);
      });

      it('[REVERSE] can find a path from usdc => eth', () => {
        const eth = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH')!;
        const usdc = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC')!;
        const path = oracles.findPath(usdc.id, eth.id, Network.ArbitrumOne);
        expect(path).toEqual([usdc.id, eth.id]);

        const value = oracles.getLatestFromPath(Network.ArbitrumOne, path);
        expect(oracles.formatNumber(oracles.invertRate(value!))).toBeCloseTo(
          ETH_PRICE
        );
      });

      it('[MULTIHOP] can find a path from wsteth => wbtc', () => {
        const wsteth = tokens.getTokenBySymbol(Network.ArbitrumOne, 'wstETH')!;
        const wbtc = tokens.getTokenBySymbol(Network.ArbitrumOne, 'WBTC')!;
        const path = oracles.findPath(wbtc.id, wsteth.id, Network.ArbitrumOne);

        const value = oracles.getLatestFromPath(Network.ArbitrumOne, path);
        expect(oracles.formatNumber(value!)).toBeCloseTo(13.253);
      });

      it('[MULTIHOP] can find a path from fusdc => fdai', () => {
        const fusdc = tokens.getTokenByID(
          Network.ArbitrumOne,
          encodeERC1155Id(3, 1695168000, AssetType.FCASH_ASSET_TYPE)
        );
        const fdai = tokens.getTokenByID(
          Network.ArbitrumOne,
          encodeERC1155Id(2, 1695168000, AssetType.FCASH_ASSET_TYPE)
        );
        const path = oracles.findPath(fusdc.id, fdai.id, Network.ArbitrumOne);

        const value = oracles.getLatestFromPath(Network.ArbitrumOne, path);
        expect(oracles.formatNumber(value!)).toBeCloseTo(0.999);
      });
    });

    describe('Risk Adjusted', () => {
      let ETH: TokenDefinition;
      let USDC: TokenDefinition;
      let FRAX: TokenDefinition;
      beforeAll(() => {
        ETH = tokens.getTokenBySymbol(Network.ArbitrumOne, 'ETH');
        USDC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'USDC');
        FRAX = tokens.getTokenBySymbol(Network.ArbitrumOne, 'FRAX');
      });

      it('ETH Asset', () => {
        const eth = TokenBalance.fromFloat(1, ETH);
        const riskAdjusted = eth.toRiskAdjustedUnderlying();
        expect(riskAdjusted.toDisplayStringWithSymbol(3)).toBe(`0.810 ETH`);
      });

      it('ETH Debt', () => {
        const eth = TokenBalance.fromFloat(-1, ETH);
        const riskAdjusted = eth.toRiskAdjustedUnderlying();
        expect(riskAdjusted.toDisplayStringWithSymbol(3)).toBe(`-1.240 ETH`);
      });

      it('Neg USDC to ETH', () => {
        const riskAdjusted = TokenBalance.fromFloat(-ETH_PRICE, USDC)
          .toRiskAdjustedUnderlying()
          .toToken(ETH);
        expect(riskAdjusted.toDisplayStringWithSymbol(3)).toBe(`-1.090 ETH`);
      });

      it('Pos USDC to ETH', () => {
        const riskAdjusted = TokenBalance.fromFloat(ETH_PRICE, USDC)
          .toRiskAdjustedUnderlying()
          .toToken(ETH);
        expect(riskAdjusted.toDisplayStringWithSymbol(3)).toBe(`0.920 ETH`);
      });

      it('Pos FRAX to ETH', () => {
        const riskAdjusted = TokenBalance.fromFloat(ETH_PRICE, FRAX)
          .toRiskAdjustedUnderlying()
          .toToken(ETH);
        expect(riskAdjusted.toDisplayStringWithSymbol(3)).toBe(`0.000 ETH`);
      });

      it('Neg fUSDC to USDC', () => {
        const fusdc = tokens.getTokenByID(
          Network.ArbitrumOne,
          encodeERC1155Id(3, 1695168000, AssetType.FCASH_ASSET_TYPE)
        );
        const usdc = TokenBalance.fromFloat(-1, fusdc);
        const riskAdjusted = usdc.toRiskAdjustedUnderlying();
        expect(riskAdjusted.toFloat()).toBeCloseTo(-1.09);
      });

      it('Pos fUSDC to USDC', () => {
        const fusdc = tokens.getTokenByID(
          Network.ArbitrumOne,
          encodeERC1155Id(3, 1695168000, AssetType.FCASH_ASSET_TYPE)
        );
        const usdc = TokenBalance.unit(fusdc);
        const riskAdjusted = usdc.toRiskAdjustedUnderlying();
        expect(riskAdjusted.toFloat()).toBeCloseTo(0.9016);
      });

      it('nUSDC to USDC', () => {
        const nUSDC = tokens.getTokenBySymbol(Network.ArbitrumOne, 'nUSDC');
        const usdc = TokenBalance.unit(nUSDC);
        const riskAdjusted = usdc.toRiskAdjustedUnderlying();
        expect(riskAdjusted.toFloat()).toBeCloseTo(0.83);
      });
    });
  }
);
