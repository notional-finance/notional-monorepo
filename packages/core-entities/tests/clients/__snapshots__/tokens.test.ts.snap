// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Token Registry Exchange Rates [MULTIHOP] can find a path from wsteth => wbtc 1`] = `14.042773512217744`;

exports[`Token Registry Exchange Rates can convert prime cash to fiat currency 1`] = `"Â¥139.40654500"`;

exports[`Token Registry Exchange Rates can note to fiat currency 1`] = `"$0.17436800"`;

exports[`Token Registry Exchange Rates can note to fiat currency 2`] = `"0.00009470 ETH"`;

exports[`Token Registry Exchange Rates can note to fiat currency 3`] = `"0.00000593 WBTC"`;

exports[`Token Registry Token Balance can convert prime cash to underlying tokens 1`] = `"1.00000119 ETH"`;
